### PR TITLE
feat: implement functionality to create credential GitHub apps

### PR DESCRIPTION
**Pull request by bahrmichael**
Closes SRCH-663

This is a follow-up to previous PRs, where we added database fields to support the new github apps integration.

See initiative "Batch Changes using GitHub App auth" on linear.

## Test plan

- Manual testing

### CODE REVIEW

File: internal/batches/service/service.go [hunk 1]
```diff
@@ -1798,3 +1812,184 @@
...
+	existing, err := s.store.UserCredentials().GetByScope(ctx, userCredentialScope)
+	if err != nil && !errcode.IsNotFound(err) {
+		return nil, err
+	}
+	if existing != nil {
+		return nil, ErrDuplicateCredential{}
+	}
+
+	a, err := s.generateAuthenticatorForCredential(ctx, generateAuthenticatorForCredentialArgs{
+		externalServiceType:    args.ExternalServiceType,
```
[the remark this answers is not in the document]
Comment (bahrmichael) [reply internal/batches/service/service.go:1857]: I moved this code from `cmd/frontend/internal/batches/resolvers/resolver.go`.

File: internal/batches/service/service.go [hunk 2]
```diff
@@ -1798,3 +1812,184 @@
...
+
+	a, err := s.generateAuthenticatorForCredential(ctx, generateAuthenticatorForCredentialArgs{
+		externalServiceType: externalServiceType,
+		externalServiceURL:  externalServiceURL,
+		credential:          credential,
+		username:            username,
+	})
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (bahrmichael) [reply internal/batches/service/service.go:1906]: I moved this code from `cmd/frontend/internal/batches/resolvers/resolver.go`.

File: cmd/frontend/internal/batches/resolvers/credential.go [hunk 3]
```diff
@@ -72,6 +73,7 @@ func commentSSHKey(ssh auth.AuthenticatorWithSSH) string {
 }
 
 type batchChangesUserCredentialResolver struct {
+	ghastore   ghastore.GitHubAppsStore
```
Comment (BolajiOlajide) [inline cmd/frontend/internal/batches/resolvers/credential.go:76]: This is a duplicate field. There's a field in [here on R80](https://github.com/sourcegraph/sourcegraph/pull/63635/files#diff-8febe19ec37d1bfa67bf2999a628e7f0249698869ece953173a04f64413371adR80) that has this already.

Can we consolidate them? `ghastore` and `ghstore`

File: cmd/frontend/internal/batches/resolvers/errors.go [hunk 4]
```diff
@@ -36,7 +36,7 @@ func (e ErrBatchChangesDisabledForUser) Extensions() map[string]any {
 	return map[string]any{"code": "ErrBatchChangesDisabledForUser"}
 }
 
-type ErrBatchChangeInvalidName struct{ error }
```
Comment (BolajiOlajide) [inline cmd/frontend/internal/batches/resolvers/errors.go:39]: I don't think we should remove this. I understand this field is unused but if this is wrapped by any error down the line, it'd be hard to unwrap.

File: cmd/frontend/internal/batches/resolvers/resolver.go [hunk 5]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/sourcegraph/sourcegraph/internal/batches/sources"
+	ghstore "github.com/sourcegraph/sourcegraph/internal/github_apps/store"
+	"github.com/sourcegraph/sourcegraph/lib/pointers"
```
Comment (BolajiOlajide) [inline cmd/frontend/internal/batches/resolvers/resolver.go:9]: These imports aren't properly sorted. 

It's one of the weirdness I encountered with Goland, I haven't found a way to integrate `golangci-lint`.

File: cmd/frontend/internal/batches/resolvers/resolver.go [hunk 6]
```diff
@@ -1123,145 +1124,27 @@ func (r *Resolver) CreateBatchChangesCredential(ctx context.Context, args *graph
...
 
-	// Validate the newly created authenticator.
-	if err := svc.ValidateAuthenticator(ctx, externalServiceURL, externalServiceType, a); err != nil {
-		return nil, &ErrVerifyCredentialFailed{SourceErr: err}
+	cred, err := svc.CreateBatchChangesSiteCredential(ctx, args.ExternalServiceURL, extsvc.KindToType(kind), args.Credential, args.Username, sources.AuthenticationStrategyUserCredential)
+	if err != nil {
+		return nil, err
 	}
-	return a, nil
+	return &batchChangesSiteCredentialResolver{credential: cred}, nil
```
Comment (BolajiOlajide) [inline cmd/frontend/internal/batches/resolvers/resolver.go:1147]: We also need to pass in the `GitHubAppsStore`

File: cmd/frontend/internal/githubapp/httpapi.go [hunk 7]
```diff
@@ -190,11 +212,13 @@ func (srv *gitHubAppServer) newAppStateHandler(w http.ResponseWriter, r *http.Re
 		http.Error(w, fmt.Sprintf("Unexpected error when generating state parameter: %s", err.Error()), http.StatusInternalServerError)
 		return
 	}
```
Comment (BolajiOlajide) [inline cmd/frontend/internal/githubapp/httpapi.go:214]: Are there plans to fix this TODO in this PR?

File: cmd/frontend/internal/githubapp/httpapi.go [hunk 8]
```diff
@@ -417,12 +463,42 @@ func (srv *gitHubAppServer) setupHandler(w http.ResponseWriter, r *http.Request)
...
 		})
 		if err != nil {
-			redirectURL := generateRedirectURL(&stateDetails.Domain, &installationID, &stateDetails.AppID, &app.Name, errors.Newf("Unexpected error while creating GitHub App installation: %s", err.Error()))
+			redirectURL := generateRedirectURL(&stateDetails.Domain, &installationID, &stateDetails.AppID, &app.Name, app.Kind, errors.Newf("Unexpected error while creating GitHub App installation: %s", err.Error()))
 			http.Redirect(w, r, redirectURL, http.StatusFound)
 			return
 		}
 
-		redirectURL := generateRedirectURL(&stateDetails.Domain, &installationID, &app.ID, &app.Name, nil)
+		if kind == ghtypes.UserCredentialGitHubAppKind {
```
Comment (BolajiOlajide) [inline cmd/frontend/internal/githubapp/httpapi.go:471]: You'll need to replicate this for Site Credentials, this only works for user credentials right now.

File: internal/batches/service/service.go [hunk 9]
```diff
@@ -1798,3 +1812,184 @@ func (s *Service) GetChangesetsByIDs(ctx context.Context, batchChange int64, ids
...
+type generateAuthenticatorForCredentialArgs struct {
+	externalServiceType    string
+	externalServiceURL     string
+	credential             string
+	username               *string
+	authenticationStrategy sources.AuthenticationStrategy
+	gitHubAppKind          ghtypes.GitHubAppKind
+}
+
+func (s *Service) generateAuthenticatorForCredential(ctx context.Context, args generateAuthenticatorForCredentialArgs) (extsvcauth.Authenticator, error) {
```
Comment (BolajiOlajide) [inline internal/batches/service/service.go:1927]: I didn't handle this earlier but we'll need a check for if `authenticationStrategy` is GitHubApp so we can create a github app authenticator.

File: internal/batches/sources/sources.go [hunk 10]
```diff
@@ -228,13 +229,23 @@ func (s *sourcer) ForUser(ctx context.Context, tx SourcerStore, uid int32, repo
...
-	// Empty authenticators are not allowed.
+type ForExternalServiceOpts struct {
+	ExternalServiceID   string
+	ExternalServiceType string
+	GitHubAppAccount    string
+	GitHubAppKind       ghtypes.GitHubAppKind
+}
+
+func (s *sourcer) ForExternalService(ctx context.Context, tx SourcerStore, au auth.Authenticator, opts ForExternalServiceOpts, as AuthenticationStrategy) (ChangesetSource, error) {
+	// Empty authenticators are not for non-GitHubApp authentication strategies.
```
Comment (BolajiOlajide) [inline internal/batches/sources/sources.go:240]: This comment is redundant, we should remove it. It should just read `// Empty authenticators are not allowed.`

File: internal/github_apps/store/store.go [hunk 11]
```diff
@@ -70,7 +70,7 @@ type GitHubAppsStore interface {
 	GetBySlug(ctx context.Context, slug string, baseURL string) (*ghtypes.GitHubApp, error)
 
 	// GetByDomain retrieves a GitHub App from the database by domain and base url
-	GetByDomain(ctx context.Context, domain itypes.GitHubAppDomain, baseURL string) (*ghtypes.GitHubApp, error)
+	GetByDomain(ctx context.Context, domain itypes.GitHubAppDomain, kind ghtypes.GitHubAppKind, baseURL string) (*ghtypes.GitHubApp, error)
```
Comment (BolajiOlajide) [inline internal/github_apps/store/store.go:73]: We should renamed this `GetByDomainAndKind` so it's clear what's going on here. The `domain` isn't the main source of truth for github apps anymore.

**Review by BolajiOlajide — CHANGES_REQUESTED** [review]

File: internal/github_apps/store/store.go (hunk 11 above)
Comment (bahrmichael) [reply internal/github_apps/store/store.go]: Done! I wanted to wait with that until I got a reasonable stable state.

File: cmd/frontend/internal/githubapp/httpapi.go (hunk 7 above)
Comment (bahrmichael) [reply cmd/frontend/internal/githubapp/httpapi.go:214]: Yes! I'll address that right away.

File: cmd/frontend/internal/batches/resolvers/errors.go (hunk 4 above)
Comment (bahrmichael) [reply cmd/frontend/internal/batches/resolvers/errors.go:39]: I'll bring it back 👍

File: cmd/frontend/internal/batches/resolvers/resolver.go (hunk 5 above)
Comment (bahrmichael) [reply cmd/frontend/internal/batches/resolvers/resolver.go:9]: I'm wondering what the source (or tool) of truth is. I just ran `sg lint --fix format` and that didn't update imports.

File: cmd/frontend/internal/batches/resolvers/resolver.go (hunk 6 above)
Comment (bahrmichael) [reply cmd/frontend/internal/batches/resolvers/resolver.go]: 👍

File: cmd/frontend/internal/batches/resolvers/credential.go (hunk 3 above)
Comment (bahrmichael) [reply cmd/frontend/internal/batches/resolvers/credential.go]: Yup, done that!

File: cmd/frontend/internal/githubapp/httpapi.go (hunk 8 above)
Comment (bahrmichael) [reply cmd/frontend/internal/githubapp/httpapi.go:476]: Done, but not tested (apart from existing CI)

File: cmd/frontend/internal/batches/resolvers/resolver.go (hunk 5 above)
Comment (BolajiOlajide) [reply cmd/frontend/internal/batches/resolvers/resolver.go:9]: Yeah, `golangci-lint` is what usually does this on vscode, but I haven't found a way to do it on GoLand.
CleanShot 2024-07-04 at 18 19 21@2x

Comment (BolajiOlajide) [reply cmd/frontend/internal/batches/resolvers/resolver.go:9]: CleanShot 2024-07-04 at 18 29 41@2x

https://plugins.jetbrains.com/plugin/12496-go-linter

I was able to get it working with this.

Comment (bahrmichael) [reply cmd/frontend/internal/batches/resolvers/resolver.go]: We should put this into our handbook, golangcli-lint doesn't seem to mentioned anywhere in sg/sg nor in notion. `sg lint` should use it, too.

File: internal/batches/service/service.go (hunk 9 above)
Comment (bahrmichael) [reply internal/batches/service/service.go:1948]: I see that you took care of this in https://github.com/sourcegraph/sourcegraph/pull/63635/commits/3aa0a9d700a2fbc7aa20763cee70503fafe6df0e. Thank you!

**Review by BolajiOlajide — APPROVED** [review]
Just a couple of comments and this should be good to go.

**Review by BolajiOlajide — APPROVED** [review]

**Pull request merged**